streams/src/test/java/org/apache/kafka/streams/integration/WindowedChangelogRetentionIntegrationTest.java

```diff
@@ -0,0 +1,293 @@
+
+        startStreams();
+
+        verifyChangelogRetentionOfWindowedStore(storeName, expectedRetention);
+    }
+
+    @Test
+    public void sessionWindowedChangelogShouldHaveRetentionOfGapIfGapLargerThanDefaultRetention() throws Exception {
+        final Duration gap = DEFAULT_RETENTION.plus(Duration.ofHours(1));
+        final Duration expectedRetention = gap;
+        runAndVerifySessionWindows(SessionWindows.with(gap), null, expectedRetention);
+    }
+
+    @Test
+    public void sessionWindowedChangelogShouldHaveDefaultRetentionIfGapLessThanDefaultRetention() throws Exception {
+        final Duration gap = DEFAULT_RETENTION.minus(Duration.ofHours(1));
+        final Duration expectedRetention = DEFAULT_RETENTION;
+        runAndVerifySessionWindows(SessionWindows.with(gap), null, expectedRetention);
+    }
+
+    @Test
+    public void sessionWindowedChangelogShouldHaveDefaultRetentionIfGapPlusGraceLessThanDefaultRetention() throws Exception {
+        final Duration grace = Duration.ofHours(1);
+        final Duration gap = DEFAULT_RETENTION.minus(grace).minus(Duration.ofHours(1));
+        final Duration expectedRetention = DEFAULT_RETENTION;
+        runAndVerifySessionWindows(SessionWindows.with(gap).grace(grace), null, expectedRetention);
+    }
+
+    @Test
+    public void sessionWindowedChangelogShouldHaveUserSpecifiedRetentionIfUserSpecifiedRetentionEvenIfLessThanDefaultRetention() throws Exception {
+        final Duration grace = Duration.ofHours(6);
+        final Duration gap = DEFAULT_RETENTION.minus(grace).minus(Duration.ofHours(1));
+        final Duration userSpecifiedRetention = gap.plus(grace).plus(Duration.ofHours(1));
+        final Duration expectedRetention = userSpecifiedRetention;
+        runAndVerifySessionWindows(SessionWindows.with(gap).grace(grace), userSpecifiedRetention, expectedRetention);
+    }
+
+    private void runAndVerifySessionWindows(final SessionWindows window,
+                                             final Duration userSpecifiedRetention,
+                                             final Duration expectedRetention) throws Exception {
+        final String storeName = "windowed-store";
+        final Serde<Windowed<String>> windowedSerde = WindowedSerdes.sessionWindowedSerdeFrom(String.class);
+        groupedStream.windowedBy(window)
+            .count(userSpecifiedRetention != null
+                ? Materialized.<String, Long, SessionStore<Bytes, byte[]>>as(storeName).withRetention(userSpecifiedRetention)
+                : Materialized.as(storeName))
+            .toStream()
+            .to(outputTopic, Produced.with(windowedSerde, Serdes.Long()));
+
+        startStreams();
+
+        verifyChangelogRetentionOfWindowedStore(storeName, expectedRetention);
+    }
+
+    @Test
+    public void joinWindowedChangelogShouldHaveRetentionOfDoubleWindowSizeIfWindowSizeLargerThanDefaultRetention() throws Exception {
```

Comment:
Should we verify for the case if user specify `before/after` explicitly, and differently?

Reply:
Good idea!